File: ios/StockbookTests/BackupTests.swift
import XCTest
import Foundation
@testable import Stockbook

/// The backup file is the only way data leaves the phone, and importing is a
/// destructive whole-database replace — so both directions are pinned down here.
/// Backup file
final class BackupTests: XCTestCase {

    private func makeStore() -> StockbookStore {
        StockbookStore(repository: InMemoryRepository())
    }

    /// A shop survives a round trip through the file
    func testRoundTrip() throws {
        let store = makeStore()
        store.setOwnerName("Khalid Al-Amri")
        let product = store.addProduct(name: "Cisa lock", stock: 12, cost: 60, price: 95)
        store.saveBill(lines: [.init(productUID: product.uid, qty: 2, price: 90)], customer: "Ahmed Contracting", paid: 100)

        let data = try BackupService.encode(store.makeBackupDocument())
        let restored = try BackupService.decode(data)

        XCTAssertEqual(restored.version, BackupDocument.currentVersion)
        XCTAssertEqual(restored.ownerName, "Khalid Al-Amri")
        XCTAssertEqual(restored.products.count, 1)
        XCTAssertEqual(restored.products.first?.uid, product.uid)
        XCTAssertEqual(restored.bills.count, 1)
        XCTAssertEqual(restored.bills.first?.paid, 100)
        XCTAssertEqual(restored.bills.first?.lines.first?.price, 90, "the charged price, not the list price")
    }

    /// Importing replaces everything rather than merging
    func testImportReplaces() throws {
        let source = makeStore()
        source.setOwnerName("Khalid Al-Amri")
        let sourceProduct = source.addProduct(name: "Cisa lock", stock: 12, cost: 60, price: 95)
        source.saveBill(lines: [.init(productUID: sourceProduct.uid, qty: 2, price: 95)], customer: "Ahmed", paid: nil)
        let document = source.makeBackupDocument()

        let destination = makeStore()
        destination.setOwnerName("Someone Else")
        destination.addProduct(name: "Padlock", stock: 3, cost: 10, price: 20)

        destination.replaceEverything(with: document)

        XCTAssertEqual(destination.products.map(\.name), ["Cisa lock"])
        XCTAssertEqual(destination.bills.count, 1)
        XCTAssertEqual(destination.settings.ownerName, "Khalid Al-Amri")
        XCTAssertTrue(destination.settings.setupCompleted)
    }

    /// Imported bills keep numbering going instead of colliding
    func testNumberingContinues() throws {
        let source = makeStore()
        let product = source.addProduct(name: "Hinge", stock: 50, cost: 3, price: 6)
        source.saveBill(lines: [.init(productUID: product.uid, qty: 1, price: 6)], customer: "A", paid: nil)
        source.saveBill(lines: [.init(productUID: product.uid, qty: 1, price: 6)], customer: "B", paid: nil)

        let destination = makeStore()
        destination.replaceEverything(with: source.makeBackupDocument())

        XCTAssertEqual(destination.settings.nextBillNumber, 3)
    }

    /// An imported file is not this phone's backup
    func testImportDoesNotCountAsBackup() throws {
        let source = makeStore()
        source.markExported()
        let document = source.makeBackupDocument()

        let destination = makeStore()
        destination.replaceEverything(with: document)

        XCTAssertNil(destination.settings.lastExportAt,
                     "the nudge stays on until this phone writes its own file")
    }

    /// Bill lines still point at their products after an import
    func testLinesResolveAfterImport() throws {
        let source = makeStore()
        let product = source.addProduct(name: "Deadbolt", stock: 10, cost: 40, price: 70)
        let bill = try XCTUnwrap(
            source.saveBill(lines: [.init(productUID: product.uid, qty: 3, price: 70)], customer: "Sami", paid: nil)
        )
        XCTAssertEqual(source.product(uid: product.uid)?.stock, 7)

        let destination = makeStore()
        destination.replaceEverything(with: source.makeBackupDocument())

        // Voiding on the new phone has to find the product by uid, which is the
        // whole reason products carry one.
        let importedBill = try XCTUnwrap(destination.bills.first { $0.number == bill.number })
        destination.void(importedBill)

        XCTAssertEqual(destination.products.first?.stock, 10)
    }

    // MARK: Validation

    /// Junk is rejected as not a Stockbook file
    func testRejectsJunk() {
        let data = Data("this is not json".utf8)
        XCTAssertThrowsError(try BackupService.decode(data)) { error in
            XCTAssertEqual(error as? BackupError, BackupError.notStockbookData)
        }
    }

    /// Valid JSON that is not a backup is rejected
    func testRejectsForeignJSON() {
        let data = Data(#"{"hello":"world"}"#.utf8)
        XCTAssertThrowsError(try BackupService.decode(data)) { error in
            XCTAssertEqual(error as? BackupError, BackupError.notStockbookData)
        }
    }

    /// A file from a newer format is rejected, not guessed at
    func testRejectsNewerVersion() {
        let data = Data(#"{"version":99,"exportedAt":"2026-08-11T00:00:00Z","ownerName":"K","currencySymbol":"SAR ","products":[],"bills":[]}"#.utf8)
        XCTAssertThrowsError(try BackupService.decode(data)) { error in
            XCTAssertEqual(error as? BackupError, BackupError.newerVersion(found: 99))
        }
    }

    /// The filename carries the export date
    func testFilename() {
        var components = DateComponents()
        components.year = 2026
        components.month = 8
        components.day = 11
        components.hour = 12
        let date = Calendar.current.date(from: components)!

        let document = BackupDocument(
            exportedAt: date,
            ownerName: "Khalid",
            currencySymbol: "SAR ",
            products: [],
            bills: []
        )

        XCTAssertEqual(document.suggestedFilename, "stockbook-2026-08-11.json")
    }

    /// The summary line reads the way Settings shows it
    func testSummary() {
        var components = DateComponents()
        components.year = 2026
        components.month = 7
        components.day = 28
        components.hour = 12
        let date = Calendar.current.date(from: components)!

        let document = BackupDocument(
            exportedAt: date,
            ownerName: "Khalid Al-Amri",
            currencySymbol: "SAR ",
            products: (0..<8).map {
                .init(uid: UUID(), name: "P\($0)", stock: 1, cost: 1, price: 2)
            },
            bills: []
        )

        XCTAssertEqual(document.summaryLine, "Khalid Al-Amri · 8 products · 0 bills · saved 28 July 2026")
    }
}

File: android/core/src/test/kotlin/com/stockbook/core/BackupTests.kt
package com.stockbook.core

import com.stockbook.core.model.Currency
import com.stockbook.core.store.DraftLine
import com.stockbook.core.store.InMemoryRepository
import com.stockbook.core.store.StockbookStore
import com.stockbook.core.text.AppLanguage
import com.stockbook.core.text.Strings
import com.stockbook.core.transfer.BackupDocument
import com.stockbook.core.transfer.BackupError
import com.stockbook.core.transfer.BackupService
import java.time.Instant
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class BackupTests {

    private val exportedAt: Instant = Instant.parse("2026-07-28T09:41:00Z")

    @Test
    fun `a document survives the round trip`() {
        val store = StockbookStore(InMemoryRepository())
        store.setOwnerName("Khalid Al-Amri")
        val product = store.addProduct("Cisa lock", 20, 60.0, 95.0)
        store.saveBill(listOf(DraftLine(product.uid, 2, 95.0)), "Ahmed", paid = 100.0)

        val document = store.makeBackupDocument(exportedAt)
        val restored = BackupService.decode(BackupService.encode(document))

        assertEquals(document, restored)
    }

    @Test
    fun `the version is checked before the shape`() {
        // A file from a future build may decode cleanly into today's structs
        // while meaning something different, and the result of getting that
        // wrong is a destructive whole-database replace.
        val json = """{"version":99,"exportedAt":"2026-08-11T00:00:00Z","ownerName":"K","currencyCode":"SAR","products":[],"bills":[]}"""
        val error = assertFailsWith<BackupError.NewerVersion> { BackupService.decode(json) }
        assertEquals(99, error.found)
    }

    @Test
    fun `anything that is not a Stockbook file is refused`() {
        assertFailsWith<BackupError.NotStockbookData> { BackupService.decode("not json at all") }
        assertFailsWith<BackupError.NotStockbookData> { BackupService.decode("""{"hello":"world"}""") }
        assertFailsWith<BackupError.NotStockbookData> { BackupService.decode("""{"version":1}""") }
    }

    @Test
    fun `the summary line reads as a sentence`() {
        val document = BackupDocument(
            exportedAt = exportedAt,
            ownerName = "Khalid Al-Amri",
            currencyCode = "SAR",
            products = (1..8).map { BackupDocument.ProductRecord("uid-$it", "P$it", 1, 1.0, 2.0) },
            bills = emptyList()
        )
        assertEquals(
            "Khalid Al-Amri · 8 products · 0 bills · saved 28 July 2026",
            document.summaryLine(Strings(AppLanguage.ENGLISH))
        )
    }

    @Test
    fun `the filename is dated and never localised`() {
        val document = BackupDocument(
            exportedAt = exportedAt,
            ownerName = "K",
            currencyCode = "SAR"
        )
        assertEquals("stockbook-2026-07-28.zip", document.suggestedFilename)
        assertTrue(document.suggestedFilename.all { it.code < 128 })
    }
}

/**
 * The file is the only way data moves between phones, and one of those phones
 * may be an iPhone. These are the properties that make that work, and none of
 * them can be checked by looking at an Android screen.
 */
class CrossPlatformBackupTests {

    /** Exactly the shape the iOS build writes: sorted keys, ISO-8601, no nulls. */
    private val fromIPhone = """
        {
          "bills" : [
            {
              "createdAt" : "2026-07-28T09:41:00Z",
              "lines" : [
                {
                  "name" : "Cisa lock",
                  "price" : 95,
                  "productUID" : "8B7F0A2E-1C4D-4E5F-9A3B-6D2E7C8F1A05",
                  "qty" : 2
                }
              ],
              "number" : 1,
              "paid" : 100,
              "total" : 190,
              "voided" : false,
              "who" : "Ahmed Contracting"
            },
            {
              "createdAt" : "2026-07-28T10:15:00Z",
              "lines" : [
                {
                  "name" : "Padlock",
                  "price" : 25,
                  "productUID" : "1A2B3C4D-5E6F-4A7B-8C9D-0E1F2A3B4C5D",
                  "qty" : 1
                }
              ],
              "number" : 2,
              "total" : 25,
              "voided" : false,
              "who" : "Sami"
            }
          ],
          "currencyCode" : "SAR",
          "customers" : [
          ],
          "exportedAt" : "2026-07-28T11:00:00Z",
          "ownerName" : "Khalid Al-Amri",
          "payments" : [
          ],
          "products" : [
            {
              "cost" : 60,
              "name" : "Cisa lock",
              "price" : 95,
              "stock" : 18,
              "uid" : "8B7F0A2E-1C4D-4E5F-9A3B-6D2E7C8F1A05"
            }
          ],
          "version" : 1
        }
    """.trimIndent()

    @Test
    fun `a backup written on an iPhone opens here`() {
        val document = BackupService.decode(fromIPhone)

        assertEquals("Khalid Al-Amri", document.ownerName)
        assertEquals(1, document.products.size)
        assertEquals(2, document.bills.size)
        // The key Swift spells `productUID`. Renaming it to fit Kotlin's
        // conventions would strand every line item on the way across.
        assertEquals("8B7F0A2E-1C4D-4E5F-9A3B-6D2E7C8F1A05", document.bills.first().lines.first().productUid)
        assertEquals(100.0, document.bills.first().paid)
        // An absent `paid` is paid in full, not zero paid.
        assertNull(document.bills[1].paid)
        // The code is what says what these numbers mean; nothing else does.
        assertEquals("SAR", document.currencyCode)
    }

    @Test
    fun `importing an iPhone backup rebuilds the shop`() {
        val store = StockbookStore(InMemoryRepository())
        store.replaceEverything(BackupService.decode(fromIPhone))

        assertEquals("Khalid Al-Amri", store.settings.ownerName)
        assertEquals(1, store.products.size)
        assertEquals(2, store.bills.size)
        assertEquals(3, store.settings.nextBillNumber, "the next bill must not reuse a number from the file")
        assertTrue(store.settings.setupCompleted)
        assertEquals(Currency.SAR, store.settings.currency)

        // The debtors list has to survive the crossing too.
        val ahmed = assertNotNull(store.customers().firstOrNull { it.name == "Ahmed Contracting" })
        assertEquals(90.0, ahmed.owed)
    }

    @Test
    fun `what we write is what an iPhone can read`() {
        val store = StockbookStore(InMemoryRepository())
        store.setOwnerName("Khalid")
        val product = store.addProduct("Cisa lock", 20, 60.0, 95.0)
        store.saveBill(listOf(DraftLine(product.uid, 2, 95.0)), "Ahmed", null)

        val text = BackupService.encode(store.makeBackupDocument(Instant.parse("2026-07-28T11:00:00Z")))

        for (key in listOf(
            "version", "exportedAt", "ownerName", "currencyCode",
            "uid", "stock", "cost", "price",
            "number", "createdAt", "total", "who", "productUID", "qty"
        )) {
            assertTrue(text.contains("\"$key\""), "the iPhone reader expects a `$key` key")
        }

        // Gone from both platforms together. A bill is edited or removed now, so
        // there is no state to carry — and a key written here that Swift no
        // longer declares is a file the iPhone would refuse outright.
        assertTrue(!text.contains("\"voided\""), "`voided` is not part of the format any more")

        // Paid in full is an *absent* key, not a null — Swift's decoder reads
        // "not present" as nil and would refuse an explicit null.
        assertTrue(!text.contains("\"paid\""), "a bill paid in full must omit `paid` entirely")
    }

    @Test
    fun `timestamps carry no fractional seconds`() {
        // Foundation's `.iso8601` strategy emits none and refuses to parse them,
        // so a millisecond here is a file the iPhone rejects.
        val store = StockbookStore(InMemoryRepository())
        val document = store.makeBackupDocument(Instant.parse("2026-07-28T11:00:00.123456Z"))

        val text = BackupService.encode(document)
        assertTrue(text.contains("2026-07-28T11:00:00Z"), text)
        assertTrue(!text.contains(".123"), "fractional seconds would be rejected on the way in")
    }
}
